e2e: ginkgo timeline

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This ensures that the failure description for a test includes all relevant information (timeouts, failures, log output) in the order in which the events occurred. This is now supported natively by Ginkgo, therefore we can remove the workarounds that we had for this before (like wrapping the fail handler and using a custom skip method).

#### Which issue(s) this PR fixes:

Related-to: https://github.com/onsi/ginkgo/issues/1069

#### Special notes for your reviewer:

TODO:
- [x] the output tests in Kubernetes still need to be adapted
- [x] ~~truncation of the failure message if it is too large~~ - not really required, dropped

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
dependencies: update to ginkgo v2.6.1, gomega v1.24.2

Ginkgo v2.5.0 adds support for a "timeline": a full description of what happened
while a specific test ran, including failures, timeouts, and log output.

Ginkgo v2.6.0 adds ReportBeforeSuite which we need for
https://github.com/kubernetes/kubernetes/issues/114313.

pohly committed:
e2e framework: avoid storing redundant failure message string

All information that we want will be written into the failure XML element's
data. We don't need the message tag and don't want it because our
tools (kettle, testgrid, spyglass) would then just concatenate the two strings.

This gets implemented for us by Ginkgo. However, truncating the failure message
is not supported there at the moment. It's unclear how important that is,
therefore this (recently added feature) gets removed.

pohly committed:
e2e: dump full report information

The Ginkgo CLI output no longer has the full output. We need the reports
generated by Ginkgo for that.

pohly committed:
e2e: fix ginkgo parameter setup

This is a fix for 104aab81a4b72004090274627f6b01246d334b86: because
the default was not set for E2E_TEST_DEBUG_TOOL, all parameters were always
also passed to the E2E suite.

That wasn't wrong for the parameters so far, but breaks when using something
like --output-dir which is only understood by the CLI.

pohly committed:
e2e: fix ginkgo-e2e.sh with no arguments

If the script was called with no arguments, it passed "${@:-}" to the suite,
which expands to one empty argument. That's not right, "${@}" should be used
instead because it expands to nothing when empty.

pohly committed:
e2e framework: stop overriding ginkgo verbosity/progress

The default settings from Ginkgo are useful enough for Kubernetes since
timeline support was added (https://github.com/onsi/ginkgo/issues/1069).

pohly committed:
e2e framework: remove injection of failures and skips into GinkgoWriter

Ginkgo's timeline support will do that for us now.

pohly commented:
/test pull-kubernetes-e2e-kind pull-kubernetes-node-e2e-containerd pull-kubernetes-e2e-gce-ubuntu-containerd

fedebongio commented:
/triage accepted

pohly committed:
e2e framework: move JUnit write function

This makes it possible to unit test it.

pohly committed:
e2e framework: verify JUnit result, not the in-memory structs

The old tests were no longer passing with Ginkgo v2.5.0. Instead of keeping the
old approach of checking recorded spec results, now the tests actually cover
what we care about most: the results recorded in JUnit.

This also gets rid of having to repeat the stack backtrace twice (once as part
of the output, once for the separate backtrace field).

logicalhan commented:
/remove-sig instrumentation

dims commented:
If you still need this PR then please rebase, if not, please close the PR

chendave commented on test/e2e/framework/internal/unittests/cleanup/cleanup_test.go:
```diff
@@ -99,46 +102,48 @@ func init() {
+< Exit [DeferCleanup (Each)] e2e - dump namespaces | framework.go:xxx <time>
+> Enter [DeferCleanup (Each)] e2e - tear down framework | framework.go:xxx <time>
+STEP: Destroying namespace "test-namespace-zzz" for this suite. - framework.go:xxx <time>
+< Exit [DeferCleanup (Each)] e2e - tear down framework | framework.go:xxx <time>
```
okay, this is what I was thinking yesterday!  nice job!

chendave commented on test/e2e/framework/test_context.go:
```diff
@@ -530,68 +530,28 @@ func AfterReadingAllFlags(t *TestContextType) {
 // needed because the full report can become too large for tools like Spyglass
 // (https://github.com/kubernetes/kubernetes/issues/111510).
 //
 // Users who want the full report can use `--junit-report`.
```
my understanding is that as long as the `--report-dir` is specified, the junit file will not be a full report.

pohly replied:
Correct.

pohly replied:
Let me be more precise: the `junit_01.xml` in `--report-dir` will always be the truncated one. The file written via `--junit-report` will be the complete one.

chendave replied:
agreed, this is guarded by the checking of `TestContext.ReportDir`.

chendave commented on test/e2e/framework/internal/output/output.go:
```diff
@@ -17,84 +17,122 @@ limitations under the License.
+	require.NoError(t, err)
+
+	if len(actual.TestSuites) != 1 {
+		t.Fatalf("expected one test suite, got %d:\n%s", len(actual.TestSuites), string(data))
```
nit: t.Fatalf("expected one test suite, got %d, junit file: %s", len(actual.TestSuites), string(data))

pohly replied:
Added "JUnit content".

chendave commented on test/e2e/framework/internal/junit/junit.go:
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
```
Does this file is actually tested anywhere? if not, why not leave it until the UT is possible?

pohly replied:
This is an oversight: output.go was meant to call it instead of using a copy. Fixed.

chendave reviewed (commented):
LGTM, just some questions or nits.

pohly commented on hack/ginkgo-e2e.sh:
```diff
@@ -183,6 +195,19 @@ case "${E2E_TEST_DEBUG_TOOL:-ginkgo}" in
+        # - JSON because it is a faithful representation of
+        #   all available information.
+        #
+        # This has go be passed to the CLI, the suite doesn't support --output-dir.
```
Fixed.

pohly commented on test/e2e/framework/test_context.go:
```diff
@@ -549,8 +549,8 @@ func writeJUnitReport(report ginkgo.Report) {
-		// the test output captured in build-log.txt has all of this
-		// information, so we don't need it in the XML.
+		// jobs that use ginkgo-e2e.sh will dump the full
+		// information into ginkgo_report.xml/json.
```
Not worth fixing the commit, this comment gets removed later...

pohly commented on hack/ginkgo-e2e.sh:
```diff
@@ -183,6 +195,19 @@ case "${E2E_TEST_DEBUG_TOOL:-ginkgo}" in
+        #   all available information.
+        #
+        # This has go be passed to the CLI, the suite doesn't support --output-dir.
+        program+=("--output-dir=${report_dir}" "--junit-report=ginkgo_report.xml" "--json-report=ginkgo_report.json")
```
We are passing the option to `ginkgo` (= the CLI), therefore `--junit-report` is the right choice.

But perhaps I misunderstand this comment.

chendave replied:
I just meant that it's possible to call this script `ginkgo-e2e.sh` with the junit file name specified as `--ginkgo.junit-report=xxx`, now the filename is renamed from `xxx` to `ginkgo_report.xml`.

pohly replied:
Passing `--ginkgo.junit-report=xxx` still works because it is passed to e2e.test after the `--ginkgo.junit-report` argument that the CLI itself adds. However, then the default `ginkgo_report.xml` is empty because the CLI doesn't get the individual XML files where it expects them. I think that's reasonable because the user gets what they explicitly asked for.

pohly commented on test/e2e/framework/log.go:
```diff
@@ -45,61 +41,9 @@ func Logf(format string, args ...interface{}) {
 	skip := 1
-	log("FAIL", "%s\n\nFull Stack Trace\n%s", msg, PrunedStack(skip))
-	ginkgo.Fail(nowStamp()+": "+msg, skip)
+	ginkgo.Fail(msg, skip)
```
I don't understand. What do you want me to change?

chendave replied:
you have an global alias defined: `var Fail = ginkgo.Fail` in this file.

pohly replied:
That's for backwards compatibility. I think calling `ginkgo.Fail` directly here is clearer.

pohly commented:
I fixed commits in https://github.com/kubernetes/kubernetes/compare/63dc239c64850e8e8c3568e9861a86b794e894ec..f63ac1dc3320d370d88bdfb1e84ede2c178358b5
and then had to rebase.

chendave commented on hack/ginkgo-e2e.sh:
```diff
@@ -230,4 +230,4 @@ case "${GINKGO_SHOW_COMMAND:-${CI:-no}}" in y|yes|true) set -x ;; esac
   ${E2E_REPORT_PREFIX:+"--report-prefix=${E2E_REPORT_PREFIX}"} \
   "${suite_args[@]:+${suite_args[@]}}" \
-  "${@:-}"
+  "${@}"
```
I confess that don't know much difference between "${@:-}" and "${@}" here, so I checked with the test script, the result shows that there will no empty parameter appended, there is no difference as far as I can see. But yes, I might be wrong.

pohly replied:
This here shows the difference:
```
$ cat /tmp/callee.sh 
#!/bin/sh

echo Number of arguments: "$#"

$ cat /tmp/caller.sh 
#!/bin/sh

echo 'With "$@":'
/tmp/callee.sh "$@"

echo 'With "${@:-}":'
/tmp/callee.sh "${@:-}"

$ /tmp/caller.sh 
With "$@":
Number of arguments: 0
With "${@:-}":
Number of arguments: 1
```

chendave replied:
luckily this haven't caused any problem so far.

chendave commented on test/e2e/framework/test_context.go:
```diff
@@ -526,72 +522,23 @@ func AfterReadingAllFlags(t *TestContextType) {
+			// for tools like Spyglass
+			// (https://github.com/kubernetes/kubernetes/issues/111510).
+			//
+			// Users who want the full report can use `--junit-report`.
```
those comments seems like dup with https://github.com/kubernetes/kubernetes/pull/113538/files#diff-daa6e9e53080eff01e28fece2f47c94a37d88d98717c54a0e8ab3d488d1aaaa4R25-R28 and could be removed from here, btw, I am fine with it.

pohly replied:
Those are two different items that get documented. I think it's better to repeat some information than to assume that the reader remembers a comment that is made elsewhere.

pohly committed:
ginkgo-e2e.sh: remove deprecated --slow-spec-threshold

From the warning message that ginkgo now emits:

  --slow-spec-threshold is deprecated --slow-spec-threshold has been deprecated
  and will be removed in a future version of Ginkgo.  This feature has proved
  to be more noisy than useful.  You can use --poll-progress-after, instead, to
  get more actionable feedback about potentially slow specs and understand
  where they might be getting stuck.

We already use --poll-progress-after.

pohly committed:
e2e: remove deprecated SuppressProgressReporting

It's not needed anymore and ginkgo warns about it.

pohly commented:
I noticed some deprecation warnings from ginkgo that I missed earlier and added two commits to address that.

chendave reviewed (approved):
Not sure if you want to bump to 2.6.1 lately, but this all lgtm.

/lgtm

pohly commented:
Yes, let's use the latest ginkgo and gomega - both updated.

@aojea : when do you want to merge this? I'm still around this week, then back in the new year.

pohly commented:
/assign @dims 

For deps review.

aojea commented:
changing reporting things use to have consequences on tools that process those .... I prefer to wait for new year 

Can you get someone to review all the bashism?

pohly commented:
> changing reporting things use to have consequences on tools that process those .... I prefer to wait for new year

Fair enough.

> Can you get someone to review all the bashism?

@chendave reviewed the script changes. @chendave, can you re-add your LGTM?

bart0sh commented:
/priority important-soon

liggitt commented:
lgtm from a dependency perspective, no new dependencies, made our graph a little shallower 🎉 

```
 	"directDependencies": 212,
 	"transitiveDependencies": 234,
 	"totalDependencies": 285,
-	"maxDepthOfDependencies": 26
+	"maxDepthOfDependencies": 23
 }
```

can tag once the gnarly test changes have review/approve :)

dims commented:
> lgtm from a dependency perspective, no new dependencies, made our graph a little shallower 🎉

LOVE IT!

aojea commented on hack/ginkgo-e2e.sh:
```diff
@@ -163,6 +163,18 @@ if [[ "${GINKGO_NO_COLOR}" == "y" ]]; then
   ginkgo_args+=("--no-color")
 fi
 
+if [[ "${E2E_REPORT_DIR:-}" ]]; then
```
don't you use `[[ -n "${E2E_REPORT_DIR:-}" ]] ?

pohly replied:
Both are equivalent. Let me add  `-n` if that is more familiar => https://github.com/kubernetes/kubernetes/compare/b7e6ff2dacdb4c1bf49f82cc9d810af5250e4052..cbb6ea5ca462b9de664d0eba9a0495847537408c

aojea commented on hack/ginkgo-e2e.sh:
```diff
@@ -183,6 +195,19 @@ case "${E2E_TEST_DEBUG_TOOL:-ginkgo}" in
     fi
     program+=("${ginkgo_args[@]:+${ginkgo_args[@]}}")
+
+    if [[ "${report_dir:-}" ]]; then
```
`if [[ -n "${report_dir:-}" ]]; then `?

aojea commented:
@pohly it was easier to review than I expected, 2 comments to the bash script and LGTM, if no voices against we can merge tomorrow and have still 2 days before christmas, WDYT?

pohly commented:
> if no voices against we can merge tomorrow and have still 2 days before christmas, WDYT?

Works for me.

pohly commented:
>  it was easier to review than I expected

Not all of my PRs are as large as PR #112923 :sweat_smile:

aojea commented:
/lgtm
/approve

@dims or @liggitt for pending approvals

Thanks

aojea commented:
checked 2 jobs, the new report files are about ~50Mb

dims commented:
/lgtm
/approve

thanks @aojea and @pohly

pohly commented:
@liggitt: can you complete the approval? 

staging/src/k8s.io/api is missing (go.mod/sum changes).

liggitt commented:
/approve

pohly commented:
/retest

BenTheElder reviewed (commented):
submitting very belated comment for the record, while discussing https://kubernetes.slack.com/archives/C7J9RP96G/p1675982155008399 found that I had a pending comment still ...

pohly commented on hack/ginkgo-e2e.sh:
```diff
@@ -163,6 +163,18 @@ if [[ "${GINKGO_NO_COLOR}" == "y" ]]; then
+if [[ "${E2E_REPORT_DIR:-}" ]]; then
+    report_dir="${E2E_REPORT_DIR}"
+else
+    # Some jobs don't use E2E_REPORT_DIR and instead pass --report-dir=<dir>
```
With the current approach the script is passing additional parameters to the ginkgo CLI and thus needs to know the report dir.

Perhaps it is possible to not use the CLI for it and instead write full reports from inside the `e2e.test` binary  - let me try that.

pohly replied:
See https://github.com/kubernetes/kubernetes/pull/115678
